src/proto/uprotocol/uuri.rs

```diff
@@ -11,38 +11,617 @@
+            resource: resource.into(),
+            ..Default::default()
+        })
     }
 }
 
 impl TryFrom<&str> for UUri {
     type Error = SerializationError;
 
-    fn try_from(value: &str) -> Result<Self, Self::Error> {
-        LongUriSerializer::deserialize(value.into())
+    /// Attempts to serialize a `String` into a `UUri`.
+    ///
+    /// # Arguments
+    ///
+    /// * `uri` - The `String` to be converted into a `UUri`.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing either the `UUri` representation of the URI or a `SerializationError`.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use up_rust::uprotocol::{UAuthority, UEntity, UResource};
+    /// use up_rust::uprotocol::UUri;
+    ///
+    /// let uri = UUri {
+    ///     entity: Some(UEntity {
+    ///         name: "example.com".to_string(),
+    ///         ..Default::default()
+    ///     })
+    ///     .into(),
+    ///     resource: Some(UResource {
+    ///         name: "rpc".to_string(),
+    ///         instance: Some("raise".to_string()),
+    ///         ..Default::default()
+    ///     })
+    ///     .into(),
+    ///     authority: None.into(),
+    ///     ..Default::default()
+    /// };
+    ///
+    /// let uri_from = UUri::try_from("/example.com//rpc.raise").unwrap();
+    /// assert_eq!(uri, uri_from);
+    /// ````
+    fn try_from(uri: &str) -> Result<Self, Self::Error> {
+        UUri::from_str(uri)
     }
 }
 
 impl TryFrom<UUri> for Vec<u8> {
     type Error = SerializationError;
 
     fn try_from(value: UUri) -> Result<Self, Self::Error> {
-        MicroUriSerializer::serialize(&value)
+        MicroUriSerializer::serialize(&value).map_err(|e| SerializationError::new(e.to_string()))
     }
 }
 
 impl TryFrom<Vec<u8>> for UUri {
     type Error = SerializationError;
 
     fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
-        MicroUriSerializer::deserialize(value)
+        MicroUriSerializer::deserialize(value).map_err(|e| SerializationError::new(e.to_string()))
+    }
+}
+
+impl UUri {
+    /// Builds a fully resolved `UUri` from the serialized long format and the serialized micro format.
+    ///
+    /// # Arguments
+    /// * `long_uri` - uri serialized as a string.
+    /// * `micro_uri` - uri serialized as a byte slice.
+    ///
+    /// # Returns
+    /// If successful, returns an UUri object serialized from the input formats. Returns `SerializationError` if the deserialization
+    /// fails or the resulting uri cannot be resolved.
+    pub fn build_resolved(long_uri: &str, micro_uri: &[u8]) -> Result<UUri, SerializationError> {
+        if long_uri.is_empty() && micro_uri.is_empty() {
+            return Err(SerializationError::new("Input uris are empty"));
+        }
+
+        let long_uri = UUri::from_str(long_uri)?;
+        let micro_uri = UUri::try_from(micro_uri.to_vec())?;
+
+        let mut auth = micro_uri.authority.unwrap_or_default();
```

Comment:
I suppose the idea here is that we would have failed and bailed out due to the two lines prior where we attempt to deserialize, so it's safe to assume that we have the appropriate values available for these three lines. Is that right?

If so, I wouldn't mind some comment to that effect, but WDYT?

Reply:
I guess in general I get a little uncomfortable pulling out the `unwrap_or_default()` unless there's clear preconditions that prevent us from _actually picking the default_.

What about using `ok_or_else` something like this?
```rust
let mut auth = micro_uri.authority.ok_or_else(|| SerializationError::new("Missing UAuthority");
let mut ue = micro_uri.entity.ok_or_else(|| SerializationError::new("Missing UEntity");
let mut ure = long_uri.resource.ok_or_else(|| SerializationError::new("Missing UResource");
```
It'd make me a little more comfortable if there's some refactoring in the future to guard against that possibility. Or maybe I'm being paranoid, haha.

WDYT?

Reply:
Valid points - added more error reporting